### PR TITLE
[spaceship] Improve Logger

### PR DESCRIPTION
**Pull request by janpio**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
`spaceship` logs each request it sends in `/tmp`. These logfiles are the easiest way to debug what spaceship is doing, and to find out why something is not working as it should.
 
Unfortunately that logging was not complete or even broken (many URLs were missing, POST body was missing, log severity was not output, the http response code was not logged and failing requests were even not logged at all) - which can be a problem like in https://github.com/fastlane/fastlane/issues/13666#issuecomment-441075575 :/

### Description
This PR makes sure that _all_ requests from spaceship are logged, include an URL all the time and a body if available (as this includes the login password, I overwrote it with `***`). I changed the log format to include severity and http status code.
 
A logfile of `fastlane spaceauth` in `master`:
```
# Logfile created on 2018-11-22 22:51:47 +0100 by logger.rb/56815
[22:51:47]: >> POST:  
[22:51:47]: << POST: : {"authType"=>"hsa2"}
[22:51:47]: >> GET:  
[22:51:48]: << GET: : {"trustedPhoneNumbers"=>[{"numberWithDialCode"=>"+49 ••••• •••••81", "pushMode"=>"sms", "obfuscatedNumber"=>"••••• •••••81", "id"=>2}, {"numberWithDialCode"=>"+49 •••• •••••85", "pushMode"=>"sms", "obfuscatedNumber"=>"•••• •••••85", "id"=>1}], "securityCode"=>{"length"=>6, "tooManyCodesSent"=>false, "tooManyCodesValidated"=>false, "securityCodeLocked"=>false}, "authenticationType"=>"hsa2", "recoveryUrl"=>"https://iforgot.apple.com/phone/add?prs_account_nm=piotrowski@gmail.com&autoSubmitAccount=true&appId=142", "cantUsePhoneNumberUrl"=>"https://iforgot.apple.com/iforgot/phone/add?context=cantuse&prs_account_nm=piotrowski@gmail.com&autoSubmitAccount=true&appId=142", "recoveryWebUrl"=>"https://iforgot.apple.com/password/verify/appleid?prs_account_nm=piotrowski@gmail.com&autoSubmitAccount=true&appId=142", "repairPhoneNumberUrl"=>"https://gsa.apple.com/appleid/account/manage/repair/verify/phone", "repairPhoneNumberWebUrl"=>"https://appleid.apple.com/widget/account/repair?#!repair", "aboutTwoFactorAuthenticationUrl"=>"https://support.apple.com/kb/HT204921", "autoVerified"=>false, "showAutoVerificationUI"=>false, "managedAccount"=>false, "supportsRecovery"=>true, "hsa2Account"=>true, "trustedPhoneNumber"=>{"numberWithDialCode"=>"+49 ••••• •••••81", "pushMode"=>"sms", "obfuscatedNumber"=>"••••• •••••81", "id"=>2}}
[22:51:52]: >> POST:  
[22:51:53]: << POST: : 
[22:51:53]: >> GET:  
[22:51:54]: << GET: : 
[22:51:54]: >> GET: https://olympus.itunes.apple.com/v1/session 
[22:51:55]: << GET: https://olympus.itunes.apple.com/v1/session: {"user"=>{"fullName"=>"Jan Piotrowski", "emailAddress"=>"piotrowski@gmail.com", "prsId"=>"1326870224"}, "provider"=>{"providerId"=>424253, "name"=>"Jan Piotrowski", "contentTypes"=>["SOFTWARE"], "subType"=>"INDIVIDUAL"}, "availableProviders"=>[{"providerId"=>424253, "name"=>"Jan Piotrowski", "contentTypes"=>["SOFTWARE"], "subType"=>"INDIVIDUAL"}], "backingType"=>"ITC", "backingTypes"=>["ITC"], "roles"=>["ADMIN", "LEGAL"], "unverifiedRoles"=>[], "featureFlags"=>["apiKeys"], "agreeToTerms"=>true, "termsSignatures"=>["ITC"], "modules"=>[{"key"=>"Apps", "name"=>"ITC.HomePage.Apps.IconText", "localizedName"=>"My Apps", "url"=>"https://appstoreconnect.apple.com/WebObjects/iTunesConnect.woa/ra/ng/app", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/Apps@2x.d3ce493e56172e92aed6.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"AppAnalytics", "name"=>"ITC.HomePage.AppAnalytics.IconText", "localizedName"=>"App Analytics", "url"=>"https://analytics.itunes.apple.com/", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/AppAnalytics@2x.e19f711d943cb42d65b2.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"SalesTrends", "name"=>"ITC.HomePage.SalesTrends.IconText", "localizedName"=>"Sales and Trends", "url"=>"https://reportingitc2.apple.com/?", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/SalesTrends@2x.b1f802112426525d990a.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"FinancialReports", "name"=>"ITC.HomePage.FinancialReports.IconText", "localizedName"=>"Payments and Financial Reports", "url"=>"https://appstoreconnect.apple.com/itc/payments_and_financial_reports", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/FinancialReports@2x.a7b266a5136cc65c643b.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"Account", "name"=>"ITC.HomePage.Account.IconText", "localizedName"=>"Users and Access", "url"=>"https://appstoreconnect.apple.com/access/users", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/ManageUsers@2x.81511f3933fb2fb4b20d.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"ContractsTaxBanking", "name"=>"ITC.HomePage.ContractsTaxBanking.IconText", "localizedName"=>"Agreements, Tax, and Banking", "url"=>"https://appstoreconnect.apple.com/WebObjects/iTunesConnect.woa/da/jumpTo?page=contracts", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/ContractsTaxBanking@2x.74466eb8570dd797602e.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"Resources", "name"=>"ITC.HomePage.Resources.IconText", "localizedName"=>"Resources and Help", "url"=>"https://itunespartner.apple.com/", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/Resources@2x.3c8d0d8c08e876cf9470.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}], "helpLinks"=>[{"key"=>"All", "url"=>"https://itunespartner.apple.com", "localizedText"=>"All Resources and Help"}, {"key"=>"News", "url"=>"https://itunespartner.apple.com/news/", "localizedText"=>"News"}, {"key"=>"Videos", "url"=>"https://itunespartner.apple.com/videos/", "localizedText"=>"Videos"}, {"key"=>"Guides", "url"=>"https://itunespartner.apple.com/guides/", "localizedText"=>"Guides"}, {"key"=>"FAQ", "url"=>"https://itunespartner.apple.com/faq/", "localizedText"=>"FAQ"}, {"key"=>"ContactUs", "url"=>"https://www.apple.com/itunes/go/itunesconnect/contactus", "localizedText"=>"Contact Us"}], "localizations"=>{"signIn"=>"Sign In", "personalDetails"=>"Edit Profile", "signOut"=>"Sign Out"}, "pccDto"=>nil, "publicUserId"=>"69a495f6-0104-5733-e053-5b8c7c1155b0", "accountHolder"=>nil}
```

And the same requests logged for this branch:
```
# Logfile created on 2018-11-22 22:59:26 +0100 by logger.rb/56815
INFO  [22:59:26]: >> POST https://idmsa.apple.com/appleauth/auth/signin: {"accountName":"piotrowski@gmail.com","password":"***","rememberMe":true}
DEBUG [22:59:27]: << POST https://idmsa.apple.com/appleauth/auth/signin: 409 {"authType"=>"hsa2"}
INFO  [22:59:27]: >> GET https://idmsa.apple.com/appleauth/auth: null
DEBUG [22:59:28]: << GET https://idmsa.apple.com/appleauth/auth: 200 {"trustedPhoneNumbers"=>[{"numberWithDialCode"=>"+49 ••••• •••••81", "pushMode"=>"sms", "obfuscatedNumber"=>"••••• •••••81", "id"=>2}, {"numberWithDialCode"=>"+49 •••• •••••85", "pushMode"=>"sms", "obfuscatedNumber"=>"•••• •••••85", "id"=>1}], "securityCode"=>{"length"=>6, "tooManyCodesSent"=>false, "tooManyCodesValidated"=>false, "securityCodeLocked"=>false}, "authenticationType"=>"hsa2", "recoveryUrl"=>"https://iforgot.apple.com/phone/add?prs_account_nm=piotrowski@gmail.com&autoSubmitAccount=true&appId=142", "cantUsePhoneNumberUrl"=>"https://iforgot.apple.com/iforgot/phone/add?context=cantuse&prs_account_nm=piotrowski@gmail.com&autoSubmitAccount=true&appId=142", "recoveryWebUrl"=>"https://iforgot.apple.com/password/verify/appleid?prs_account_nm=piotrowski@gmail.com&autoSubmitAccount=true&appId=142", "repairPhoneNumberUrl"=>"https://gsa.apple.com/appleid/account/manage/repair/verify/phone", "repairPhoneNumberWebUrl"=>"https://appleid.apple.com/widget/account/repair?#!repair", "aboutTwoFactorAuthenticationUrl"=>"https://support.apple.com/kb/HT204921", "autoVerified"=>false, "showAutoVerificationUI"=>false, "managedAccount"=>false, "supportsRecovery"=>true, "hsa2Account"=>true, "trustedPhoneNumber"=>{"numberWithDialCode"=>"+49 ••••• •••••81", "pushMode"=>"sms", "obfuscatedNumber"=>"••••• •••••81", "id"=>2}}
INFO  [22:59:38]: >> POST https://idmsa.apple.com/appleauth/auth/verify/trusteddevice/securitycode: {"securityCode":{"code":"246313"}}
DEBUG [22:59:39]: << POST https://idmsa.apple.com/appleauth/auth/verify/trusteddevice/securitycode: 204 
INFO  [22:59:39]: >> GET https://idmsa.apple.com/appleauth/auth/2sv/trust: null
DEBUG [22:59:40]: << GET https://idmsa.apple.com/appleauth/auth/2sv/trust: 204 
INFO  [22:59:40]: >> GET https://olympus.itunes.apple.com/v1/session: null
DEBUG [22:59:41]: << GET https://olympus.itunes.apple.com/v1/session: 200 {"user"=>{"fullName"=>"Jan Piotrowski", "emailAddress"=>"piotrowski@gmail.com", "prsId"=>"1326870224"}, "provider"=>{"providerId"=>424253, "name"=>"Jan Piotrowski", "contentTypes"=>["SOFTWARE"], "subType"=>"INDIVIDUAL"}, "availableProviders"=>[{"providerId"=>424253, "name"=>"Jan Piotrowski", "contentTypes"=>["SOFTWARE"], "subType"=>"INDIVIDUAL"}], "backingType"=>"ITC", "backingTypes"=>["ITC"], "roles"=>["ADMIN", "LEGAL"], "unverifiedRoles"=>[], "featureFlags"=>["apiKeys"], "agreeToTerms"=>true, "termsSignatures"=>["ITC"], "modules"=>[{"key"=>"Apps", "name"=>"ITC.HomePage.Apps.IconText", "localizedName"=>"My Apps", "url"=>"https://appstoreconnect.apple.com/WebObjects/iTunesConnect.woa/ra/ng/app", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/Apps@2x.d3ce493e56172e92aed6.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"AppAnalytics", "name"=>"ITC.HomePage.AppAnalytics.IconText", "localizedName"=>"App Analytics", "url"=>"https://analytics.itunes.apple.com/", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/AppAnalytics@2x.e19f711d943cb42d65b2.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"SalesTrends", "name"=>"ITC.HomePage.SalesTrends.IconText", "localizedName"=>"Sales and Trends", "url"=>"https://reportingitc2.apple.com/?", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/SalesTrends@2x.b1f802112426525d990a.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"FinancialReports", "name"=>"ITC.HomePage.FinancialReports.IconText", "localizedName"=>"Payments and Financial Reports", "url"=>"https://appstoreconnect.apple.com/itc/payments_and_financial_reports", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/FinancialReports@2x.a7b266a5136cc65c643b.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"Account", "name"=>"ITC.HomePage.Account.IconText", "localizedName"=>"Users and Access", "url"=>"https://appstoreconnect.apple.com/access/users", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/ManageUsers@2x.81511f3933fb2fb4b20d.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"ContractsTaxBanking", "name"=>"ITC.HomePage.ContractsTaxBanking.IconText", "localizedName"=>"Agreements, Tax, and Banking", "url"=>"https://appstoreconnect.apple.com/WebObjects/iTunesConnect.woa/da/jumpTo?page=contracts", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/ContractsTaxBanking@2x.74466eb8570dd797602e.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}, {"key"=>"Resources", "name"=>"ITC.HomePage.Resources.IconText", "localizedName"=>"Resources and Help", "url"=>"https://itunespartner.apple.com/", "iconUrl"=>"https://appstoreconnect.apple.com/static/img/ico_homepage/themed/apps/Resources@2x.3c8d0d8c08e876cf9470.png", "down"=>false, "visible"=>true, "hasNotifications"=>false}], "helpLinks"=>[{"key"=>"All", "url"=>"https://itunespartner.apple.com", "localizedText"=>"All Resources and Help"}, {"key"=>"News", "url"=>"https://itunespartner.apple.com/news/", "localizedText"=>"News"}, {"key"=>"Videos", "url"=>"https://itunespartner.apple.com/videos/", "localizedText"=>"Videos"}, {"key"=>"Guides", "url"=>"https://itunespartner.apple.com/guides/", "localizedText"=>"Guides"}, {"key"=>"FAQ", "url"=>"https://itunespartner.apple.com/faq/", "localizedText"=>"FAQ"}, {"key"=>"ContactUs", "url"=>"https://www.apple.com/itunes/go/itunesconnect/contactus", "localizedText"=>"Contact Us"}], "localizations"=>{"signIn"=>"Sign In", "personalDetails"=>"Edit Profile", "signOut"=>"Sign Out"}, "pccDto"=>nil, "publicUserId"=>"69a495f6-0104-5733-e053-5b8c7c1155b0", "accountHolder"=>nil}
```
### Caution!

1. This PR includes a really terrible `extract_key_from_block` method that I wrote to get a specific "key" out of a block.  
Faraday uses blocks to supply configuration to the client/request (by calling `req.url(...)` or setting e.g. `req.headers[...] = ...`) and I had no idea how to get these values _out_ of there. This is required as the URL of the request is only sometimes supplied as a proper param, but very often only in this block. And as logging the URLs was missing, I had to find a solution.  
Background here is that I still have problems to wrap my brain around how blocks actually works - in my native PHP they are just not a thing, and all the samples I found online are about outputting fruit names or counting stuff, so I was a bit lost on how to implement this here.  
So I ended up creating a fake Object that has the same methods and "properties" as the one Faraday seems to be using, and 3 getters with a known name I could use to extract the information I want.
How can I do this better? Pleeeeease help me to replace this shitty code.

2. This needs to be tested _really_ well. The additional code in the `log_*` methods caused several crashes after the naive implementation, just because the tests sent the data in a different format or structure than I initially expected. We have to make sure that all the other methods that are not properly tested (I am sure there are some of those for spaceship) do not cause similar crashes. 

3. Please take a look at the improved (?) log formatting. Does this work for you? Any problems? Logging HTTP request with params and body is notoriously difficult in a flat file (and not e.g. JSON), so I hope my idea here works.

### CODE REVIEW

File: spaceship/lib/spaceship/client.rb [hunk 1]
```diff
@@ -679,26 +694,61 @@ def store_csrf_tokens(response)
...
-    def log_response(method, url, response)
+    def log_response(method, url, response, headers = nil, &block)
+      url ||= extract_key_from_block('url', &block)
       body = response.body.kind_of?(String) ? response.body.force_encoding(Encoding::UTF_8) : response.body
-      logger.debug("<< #{method.upcase}: #{url}: #{body}")
+      resp_hash = response.to_hash
+      logger.debug("<< #{method.upcase} #{url}: #{resp_hash[:status]} #{body}")
+    end
+
+    def extract_key_from_block(key, &block)
```
Comment (janpio) [inline spaceship/lib/spaceship/client.rb]: This code is terrible. Really. How do I do this (the method gets a Proc in `&block`, I want to get the values from `.body`, `.headers` and `.url()`) properly?

File: spaceship/lib/spaceship/client.rb [hunk 2]
```diff
@@ -686,26 +702,78 @@ def store_csrf_tokens(response)
...
+        end
+      end
       params_to_log = Hash(params).dup # to also work with nil
       params_to_log.delete(:accountPassword) # Dev Portal
       params_to_log.delete(:theAccountPW) # iTC
       params_to_log = params_to_log.collect do |key, value|
         "{#{key}: #{value}}"
       end
-      logger.info(">> #{method.upcase}: #{url} #{params_to_log.join(', ')}")
+      logger.info(">> #{method.upcase} #{url}: #{body.to_json}#{params_to_log.join(', ')}")
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/client.rb]:
```suggestion
      logger.info(">> #{method.upcase} #{url}: #{body.to_json} #{params_to_log.join(', ')}")
```
Need a space here ^

**Review by joshdholtz — CHANGES_REQUESTED** [review]
Found a few small requests! But this looks 💯 Gunna be nice having logs like this 😊

File: spaceship/lib/spaceship/client.rb [hunk 3]
```diff
@@ -251,7 +251,8 @@ def logger
         end
 
         @logger.formatter = proc do |severity, datetime, progname, msg|
-          "[#{datetime.strftime('%H:%M:%S')}]: #{msg}\n"
+          severity = format('%-5.5s', severity)
```
[the remark this answers is not in the document]
Comment (janpio) [reply spaceship/lib/spaceship/client.rb]: You're not? Man, a _real_ developer should know...

```suggestion
          severity = format('%-5.5s', severity) # make sure `severity` is always 5 characters, filled with spaces on the right
```

(This is copy paste from Google or StackOverflow of course. Had no idea.)

File: spaceship/lib/spaceship/client.rb (hunk 2 above)
Comment (janpio) [reply spaceship/lib/spaceship/client.rb]: Indeed, thanks for noticing 🔬

File: spaceship/lib/spaceship/client.rb [hunk 4]
```diff
@@ -686,26 +702,78 @@ def store_csrf_tokens(response)
...
+            @headers
+          end
+
+          def getparams
+            @params
+          end
+        end
+        obj.headers = {}
+        yield(obj)
+        obj.public_send('get' + key)
```
[the remark this answers is not in the document]
Comment (janpio) [reply spaceship/lib/spaceship/client.rb]: Yes, from the look of it this makes sense. I got confused about the `get`-methods it seems. I remember I tried to do it with `attr_writer` and it didn't work, so let's see if I just messed up and the tests still pass.

I am still hoping for a _real_ solution to the problem by the way, where we don't have to create this franken-object. There has to be a better, proper way to access this block data somehow.

[the remark this answers is not in the document]
Comment (janpio) [reply spaceship/lib/spaceship/client.rb]: Ok, much better results with 
```
    def extract_key_from_block(key, &block)
      if block_given?
        obj = Object.new
        class << obj
          attr_accessor :body, :headers, :params, :url
          # rubocop: disable Style/TrivialAccessors
          # the block calls `url` (not `url=`) so need to define `url` method
          def url(url)
            @url = url
          end
          # rubocop: enable Style/TrivialAccessors
        end
        obj.headers = {}
        yield(obj)
        obj.instance_variable_get("@#{key}")
      end
    end
```
(Identical to your code, just replaced `attr_writer` with `attr_accessor`)

I think the `req.headers[...] = ...` uses an "accessor" first to get the full array/hash, then changes one value of it and then writes it back. Does that make sense?

**Review by joshdholtz — APPROVED** [review]
This is great! Ran through a bunch of tests and this looks 💯

**Pull request merged**